File: translated/f-to-c/DRB061-matrixvector1-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Matrix-vector multiplication: outer-level loop parallelization. No data race pairs.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

void foo()
{
    int i, j, N;
    float sum;
    float** a;
    float* v;
    float* v_out;

    N = 100;

    // Allocate arrays
    a = (float**)malloc(N * sizeof(float*));
    for (i = 0; i < N; i++) {
        a[i] = (float*)malloc(N * sizeof(float));
    }
    v = (float*)malloc(N * sizeof(float));
    v_out = (float*)malloc(N * sizeof(float));

    #pragma omp parallel for private(i, j, sum)
    for (i = 0; i < N; i++) {
        sum = 0.0;  // Initialize sum for each thread
        for (j = 0; j < N; j++) {
            sum = sum + a[i][j] * v[j];
        }
        v_out[i] = sum;
    }

    // Free arrays
    for (i = 0; i < N; i++) {
        free(a[i]);
    }
    free(a);
    free(v);
    free(v_out);
}

int main()
{
    foo();
    return 0;
} 

File: translated/f-to-go/DRB061-matrixvector1-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Matrix-vector multiplication: outer-level loop parallelization. No data race pairs.

package main

import (
	"runtime"
	"sync"
)

func foo() {
	var N int
	var a [][]float32
	var v, v_out []float32

	N = 100
	a = make([][]float32, N)
	for i := 0; i < N; i++ {
		a[i] = make([]float32, N)
	}
	v = make([]float32, N)
	v_out = make([]float32, N)

	//$omp parallel do private(i,j,sum)
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := N / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= N; start += chunkSize {
		end := start + chunkSize - 1
		if end > N {
			end = N
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				sum := float32(0.0) // sum is private to each goroutine
				for j := 1; j <= N; j++ {
					sum = sum + a[i-1][j-1]*v[j-1]
				}
				v_out[i-1] = sum // No race - each thread works on different elements
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do
}

func main() {
	foo()
}
